### PR TITLE
Fixes #11643 - [API] Add preview template to host

### PR DESCRIPTION
**Pull request by shlomizadok**

### CODE REVIEW

File: app/controllers/api/v2/hosts_controller.rb [hunk 1]
```diff
@@ -238,6 +238,18 @@ def rebuild_config
         end
       end
 
+      api :GET, "/hosts/:id/template/:type", N_("Get host template")
+      param :id, :identifier_dottable, :required => true
+      param :type, String, :desc => N_("Template type")
```
Comment (domcleal) [inline app/controllers/api/v2/hosts_controller.rb]: Is it worth including valid values in here?  You can either put an array replacing String (e.g. `TemplateKind.all...`), or inside the documentation, e.g. the  :operatingsystem_families string configured in config/initializers/apipie.rb.

Comment (domcleal) [inline app/controllers/api/v2/hosts_controller.rb]: It's probably worth calling it a template _kind_ too, which matches the other config template controller parameters.

File: app/controllers/api/v2/hosts_controller.rb [hunk 2]
```diff
@@ -238,6 +238,18 @@ def rebuild_config
         end
       end
 
+      api :GET, "/hosts/:id/template/:type", N_("Get host template")
```
Comment (domcleal) [inline app/controllers/api/v2/hosts_controller.rb]: "Preview rendered template content" or something?

File: app/controllers/api/v2/hosts_controller.rb [hunk 3]
```diff
@@ -238,6 +238,18 @@ def rebuild_config
...
+      param :id, :identifier_dottable, :required => true
+      param :type, String, :desc => N_("Template type")
+      def template
+        template = @host.provisioning_template({ :kind => params[:type] })
+        if template.nil?
+          render_message (_("No template with type %{type} for %{host}") % {:type => params[:type], :host => @host.to_label}), :status => :unprocessable_entity
```
Comment (domcleal) [inline app/controllers/api/v2/hosts_controller.rb]: I'm unsure, but is a not_found response more accurate?

Comment (domcleal) [inline app/controllers/api/v2/hosts_controller.rb]: also type/kind I think

File: app/controllers/api/v2/hosts_controller.rb (hunk 2 above)
Comment (domcleal) [inline app/controllers/api/v2/hosts_controller.rb]: To be specific, "provisioning template" too I suppose.

**Comment by domcleal** [comment]
Could you also add a functional test for the template-not-found error case please?

**Comment by shlomizadok** [comment]
%s/type/kind/g
added test if template not found

**Comment by shlomizadok** [comment]
[test]

**Comment by shlomizadok** [comment]
@domcleal, is this failing because I called the apipie substitution `template_kinds` ? Shall I rename it to `templates` ?

File: config/initializers/apipie.rb [hunk 4]
```diff
@@ -19,6 +19,7 @@
     :operatingsystem_families => Operatingsystem.families.join(", "),
     :providers => ComputeResource.providers.join(', '),
     :default_nic_type => InterfaceTypeMapper::DEFAULT_TYPE.humanized_name.downcase,
+    :template_kinds => TemplateKind.all.map(&:name).join(", ")
```
Comment (domcleal) [inline config/initializers/apipie.rb]: During Rails initialisation without a DB, this template_kinds table won't exist.  I think if the lookup was lazy and happened inside the lambda, it'd be fine.  Perhaps use a lambda here and call it within the other, or move the code so it's not in this hash?

Comment (shlomizadok) [inline config/initializers/apipie.rb]: Removed from apipie initializer and added next to the description.

**Comment by shlomizadok** [comment]
[test]

**Comment by shlomizadok** [comment]
No clue why this fails on Katello :confused:

File: app/controllers/api/v2/hosts_controller.rb [hunk 5]
```diff
@@ -238,6 +238,18 @@ def rebuild_config
         end
       end
 
+      api :GET, "/hosts/:id/template/:kind", N_("Preview rendered provisioning template content")
+      param :id, :identifier_dottable, :required => true
+      param :kind, String, :required => true, :desc => N_("Template kinds, available values: %{template_kinds}") % { :template_kinds => TemplateKind.all.map(&:name).join(", ") }
```
Comment (domcleal) [inline app/controllers/api/v2/hosts_controller.rb]: This is getting evaluated when this controller is loaded, but the TempateKind table may not be available (e.g. before db:migrate).  It needs to be lazier.

**Comment by domcleal** [comment]
This has the same problem, you're accessing the table in the class scope - if the class is evaluated during Rails initialisation (as it would be in production too, I think), then the table won't be ready if you're still DB migrating.  I think the lazy apipie method is safer.

**Comment by shlomizadok** [comment]
@domcleal, couldn't find a way to lazy load it on the apipie inititilizer. A hint?

**Comment by domcleal** [comment]
Have a look at the second commit here: https://github.com/theforeman/foreman/compare/develop...domcleal:11643-apipie-templates

Note that the apipie initialiser already sets up a lambda just below the line you're changing, so translations are already lazy.  It's the data that isn't - so I put a lambda into the data and then called it at runtime when apipie's translating the text.

**Comment by shlomizadok** [comment]
[test]

**Comment by shlomizadok** [comment]
@domcleal - you are the best

**Comment by domcleal** [comment]
Looks better, thanks.

I have a concern about the rendering method used here, which is giving different results to the preview used on the template edit page and the real templates.  It's mostly caused by instance variables again - two different sets are missing.

a) the `@provisioning_method` and OS-specific variables are missing from the template as `load_template_vars` (in the renderer) isn't called.  This is also the cause of http://projects.theforeman.org/issues/9694.  This method should be called from Host::Managed#render_template I think as that's the context in which we call unattended_render.

As an aside, these work properly in the templates controller preview, so perhaps we should also use render_template there too?

b) if you preview a PXELinux template via the API, you'll see there are no `@initrd`/`@kernel` variables set.  These are set in the TFTP orchestration's `generate_pxe_template` method and also in the unattended controller by calling `pxe_config` (in the renderer).  `pxe_config` should probably be merged into `load_template_vars` so we get it everywhere.

This is also broken in the template controller preview, so it's just working in the unattended/spoof view and during TFTP orchestration.

---

I'd quite like to see (a) fixed first so at least these two recent preview methods return consistent results, but I think (b) will also need fixing soon as users will notice that the PXELinux previews aren't working correctly compared to the unattended/spoof view.

**Comment by shlomizadok** [comment]
@domcleal - thank you for the deep review :+1: . I hope this solves a) & b)

File: lib/foreman/renderer.rb [hunk 6]
```diff
@@ -129,6 +129,7 @@ def load_template_vars
       # and is needed to direct the host to the correct url. without it, we increase
       # latency by requesting the correct url directly from the proxy.
       @template_url = params['url']
+      pxe_config
```
Comment (domcleal) [inline lib/foreman/renderer.rb]: There's a filter in UnattendedController that this can be removed from now.

Comment (domcleal) [inline lib/foreman/renderer.rb]: Er, nearly, you'll see when you look.  The load_template_vars method is only called on certain actions, but could now be called on both PROVISION and CONFIG_URLS.

**Comment by domcleal** [comment]
Untested, but that looks better, it should resolve variables properly now.  It's missing a test for these fixes though, I don't know whether you want to put a basic unit test in host_test.rb for #render_template or similar to cover it?

**Comment by shlomizadok** [comment]
@domcleal - hope this is the kind of test you were meaning to.
Also, please note that in `load_template_vars` I've conditioned pxe_config to only if the OS has pxe_type.

**Comment by domcleal** [comment]
Yes, that's OK, thanks.

There's a comment about UnattendedController inline, hidden behind: https://github.com/theforeman/foreman/pull/2889#discussion-diff-45195754.  The pxe_config call can be removed from the unattended controller filters now, and be replaced by load_template_vars.

**Comment by shlomizadok** [comment]
Added call to load_tempate_vars for `CONFIG_URLS`

**Comment by shlomizadok** [comment]
[test]

**Comment by domcleal** [comment]
Merged as 8a817cb, thanks @shlomizadok.

**Pull request closed without merging**